docs/compilers/CSharp/Nullability Public API Design Notes.md

```diff
@@ -0,0 +1,241 @@
+
+Value types are always considered `NotNull`.
+Nullable value types and reference types have their state tracked, and are either `NotNull` or `MayBeNull` depending on flow state.
+Unconstrained generic parameters are being tracked, but what they will return for flow state information before explicit null checks occur in source code has not yet been decided.
+
+### Declared Nullability
+
+Declared nullability is the nullability that was explicitly typed in the source code by the programmer.
+There are some cases where this is inferred, such as lambda arguments/return types, `var` variables, and type parameter substitution.
+For these cases, flow state is necessary to compute the declared nullability.
+For example:
+
+```C#
+string? s1 = string.Empty;
+var s2 = s1;
+s2 = null; // This will warn, because the type of s2 is string, not string?
+```
+
+The declared nullability of `s1` is `MayBeNull`, even though it is being initialized with a non-null value.
+However, because of the flow state of `s1` is `NotNull`, we will infer the declared nullability of `s2` to be `NotNull` as well.
+
+Declared nullability can be retreived from the declaration symbol (e.g. `IFieldSymbol`, `IMethodSymbol`, etc.).
+
+### Flow State
+
+When the nullable feature is enabled, Roslyn will track the flow state of expressions throughout a method, regardless of what the variable was declared as.
+`SemanticModel` can be used to request flow state information for individual expressions or sub-expressions through the `GetTypeInfo` APIs.
+The `TypeInfo` struct will be augmented with `Nullability` and `ConvertedNullability` information, similar to how it currently has `Type` and `ConvertedType`.
+Nullability of symbols at a location can be retrieved through the `GetSymbolInfo` API: if the expression successfully resolved to a single symbol, the `Symbol` property will have nullability calculated.
+`CandidateSymbols` will not have calculated nullabilities, as we cannot run nullable analysis on code that has failed overload resolution.
+
+### `GetSemanticModel(bool skipNullabilityInformation)`
+
+This API proposal changes the way `SemanticModel` calls behave by default.
+Currently, if semantic information is requested for a `SyntaxNode`, we make an effort to bind as little as possible to provide semantic information about that statement, usually a single statement.
+However, if that semantic information will include `Nullability`, we must bind all statements that come before that statement and run nullability analysis up to and including that statement.
+We believe that this change will be ok for most scenarios, as things such as colorization and analyzers are already running in the background near-constantly, binding pretty much everything on screen and in the file anyway.
+There are a few specific scenarios that will be perf-sensitive, and an overload for `GetSemanticModel()` that skips nullability information will be provided for these.
```

Comment:
this seems speculative.  why not see how far the above gets us, and add this as necessary.  in the IDE's experience, perf is already so good (with ambient features causing near-entire-file binding) that it's unclear to me if we'd need this.

Reply:
From an API standpoint, I think the evolution of `GetSemanticModel()` is the right balance (i.e. the existing API everybody uses today will just bind more stuff, but we believe perf is acceptable and that providing higher-fidelity to consumers is more important). From an API-design standpoint overloads are almost "free" in that that they have been proven to not increase the concept count for consumers, assuming the ones with fewer arguments work in the majority of cases, which seem to be the case here.

I'm not qualified to argue the perf issues with the new model and whether or not that adding the the new API is necessary or not. However, given that you can update Roslyn fairly easily, I would urge you to consider @CyrusNajmabadi proposal to exclude the API unless you have compelling evidence that the API is needed. The risk in giving an API prematurely is that by the time you understand the impacted scenario you realized that the API you have provided isn't the right one either and you need to expose a different switch (i.e. not `skipNullabilityInformation` but `skipFizBuzzAnalysis`). If the opportunity cost of doing the API later is tiny, deferring is the better approach IMHO.

Reply:
I'll add a note about this that we'll only provide this API if dogfooding shows that it's necessary.